src/framework/control.ml

```diff
@@ -595,26 +595,46 @@ struct
       let active_transformations = get_string_list "trans.activated" in
       (if active_transformations <> [] then
         (* Transformations work using Cil visitors which use the location, so we join all contexts per location. *)
+        (* TODO: would it not be better to work by CFG node here? all transformations just go stmt->location anyway *)
         let joined =
           let open Batteries in let open Enum in
           let e = LHT.enum lh |> map (Tuple2.map1 (Node.location % fst)) in (* drop context from key and get location from node *)
           let h = Hashtbl.create (if fast_count e then count e else 123) in
-          iter (fun (k,v) ->
+          iter (fun (loc, v) ->
             (* join values for the same location *)
-            let v' = try Spec.D.join (Hashtbl.find h k) v with Not_found -> v in
-            Hashtbl.replace h k v') e;
+            let v' = Hashtbl.find_option h loc |> Stdlib.Option.fold ~none:v ~some:(Spec.D.join v) in
+            Hashtbl.replace h loc v')
+            e;
           h
         in
+
+        (* Map locations of functions to their IDs, so queries can ask about functions. *)
+        let fid_by_loc = Hashtbl.create 37 in
+        LHT.enum lh
+          |> Enum.map (fst %> fst) (* grab node from key in key-value pair *)
+          |> Enum.iter (function
+              | (Function fd | FunctionEntry fd) as n -> Hashtbl.replace fid_by_loc (Node.location n) fd.svar.vid
+              | _ -> ()) ;
+
         let ask ~node loc = (fun (type a) (q: a Queries.t) ->
```

Comment:
Rewrite `ask` to use a datatype with a constructor for query, must be dead and must be uncalled.